mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+// mempool. It adds it to the rejectedTxCache so it will not be added again
+func (txmp *TxPool) RemoveTxByKey(txKey types.TxKey) error {
+	txmp.removeTxByKey(txKey)
+	txmp.metrics.EvictedTxs.Add(1)
+	return nil
+}
+
+func (txmp *TxPool) removeTxByKey(txKey types.TxKey) {
+	txmp.rejectedTxCache.Push(txKey)
+	_ = txmp.evictedTxs.Pop(txKey)
+	_ = txmp.store.remove(txKey)
+	txmp.seenByPeersSet.RemoveKey(txKey)
+	txmp.metrics.EvictedTxs.Add(1)
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	txmp.txsToBeBroadcast = make(map[types.TxKey]struct{})
+}
+
+// Flush purges the contents of the mempool and the cache, leaving both empty.
+// The current height is not modified by this operation.
+func (txmp *TxPool) Flush() {
+	// Remove all the transactions in the list explicitly, so that the sizes
+	// and indexes get updated properly.
+	size := txmp.Size()
+	txmp.store.reset()
+	txmp.seenByPeersSet.Reset()
+	txmp.evictedTxs.Reset()
+	txmp.rejectedTxCache.Reset()
+	txmp.metrics.EvictedTxs.Add(float64(size))
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	txmp.txsToBeBroadcast = make(map[types.TxKey]struct{})
+}
+
+// PeerHasTx marks that the transaction has been seen by a peer.
+func (txmp *TxPool) PeerHasTx(peer uint16, txKey types.TxKey) {
+	txmp.logger.Debug("peer has tx", "peer", peer, "txKey", fmt.Sprintf("%X", txKey))
+	txmp.seenByPeersSet.Add(txKey, peer)
+}
+
+// allEntriesSorted returns a slice of all the transactions currently in the
+// mempool, sorted in nonincreasing order by priority with ties broken by
+// increasing order of arrival time.
+func (txmp *TxPool) allEntriesSorted() []*wrappedTx {
+	txs := txmp.store.getAllTxs()
+	sort.Slice(txs, func(i, j int) bool {
+		if txs[i].priority == txs[j].priority {
+			return txs[i].timestamp.Before(txs[j].timestamp)
+		}
+		return txs[i].priority > txs[j].priority // N.B. higher priorities first
+	})
+	return txs
+}
+
+// ReapMaxBytesMaxGas returns a slice of valid transactions that fit within the
+// size and gas constraints. The results are ordered by nonincreasing priority,
+// with ties broken by increasing order of arrival. Reaping transactions does
+// not remove them from the mempool
+//
+// If maxBytes < 0, no limit is set on the total size in bytes.
+// If maxGas < 0, no limit is set on the total gas cost.
+//
+// If the mempool is empty or has no transactions fitting within the given
+// constraints, the result will also be empty.
+func (txmp *TxPool) ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs {
+	var totalGas, totalBytes int64
+
+	var keep []types.Tx //nolint:prealloc
+	for _, w := range txmp.allEntriesSorted() {
+		// N.B. When computing byte size, we need to include the overhead for
+		// encoding as protobuf to send to the application.
+		totalGas += w.gasWanted
+		totalBytes += types.ComputeProtoSizeForTxs([]types.Tx{w.tx})
+		if (maxGas >= 0 && totalGas > maxGas) || (maxBytes >= 0 && totalBytes > maxBytes) {
```

Comment:
[non-blocking] nit: I'm not totally sure how this function is used so not sure if it would be useful to devs to split this if statement into two with corresponding log statements.  My initial thought was that it would be helpful information to know if a txn you submitted was invalid due to the gas or size limits.

Reply:
More generally, one of the problems with mempools that we've discussed internally at Tendermint was how opaque it is. It's quite difficult to get solid guarantees around transactions and for users to know what's happening with their transactions. Part of it is the distributed nature of the system and part of it is the divide between consensus and state machine

Reply:
In this case, the transaction isn't invalid (that the consensus engine knows of), it just didn't fit into this block based on total gas and total byte constraints. It's perfectly possible that the transaction makes it into the next block. Also as the user, you might have submitted it to a different node to the one trying to construct the block now